scheduler/scheduler.py

```diff
@@ -898,45 +892,68 @@ def _schedule_jobs_on_workers(self):
                 self._worker_type_to_worker_id_mapping[worker_type])
             worker_state[worker_type] = {
                 'worker_ids': worker_ids,
-                'reserved_worker_ids': set(),
+                'assigned_worker_ids': set(),
                 'server_id_ptr': 0,
-                'num_workers_assigned': 0,
             }
 
-        # Keep jobs on the same server if possible.
-        already_scheduled_jobs = set()
-        new_worker_types = {}
-        for worker_type in scheduled_jobs:
-            for (job_id, scale_factor) in scheduled_jobs[worker_type]:
-                new_worker_types[job_id] = worker_type
+        prev_worker_types = {}
         for (job_id, worker_ids) in self._current_worker_assignments.items():
-            current_worker_type = \
-                self._worker_id_to_worker_type_mapping[worker_ids[0]]
-            if job_id not in new_worker_types:
-                continue
-            if new_worker_types[job_id] == current_worker_type:
-                reserved_worker_ids = \
-                    worker_state[current_worker_type]['reserved_worker_ids']
-                scale_factor = len(worker_ids)
-                for worker_id in worker_ids:
-                    reserved_worker_ids.add(worker_id)
-                already_scheduled_jobs.add(job_id)
-                self._assign_workers_to_job(job_id, scale_factor,
-                                            current_worker_type,
-                                            worker_state[current_worker_type],
-                                            new_worker_assignments,
-                                            worker_ids)
-
-        # Assign all jobs that have an updated placement.
+            worker_type = self._worker_id_to_worker_type_mapping[worker_ids[0]]
+            prev_worker_types[job_id] = worker_type
+
         for worker_type in worker_types:
-            for (job_id, scale_factor) in scheduled_jobs[worker_type]:
-                if job_id not in self._allocation:
-                    continue
-                elif job_id in already_scheduled_jobs:
-                    continue
-                self._assign_workers_to_job(job_id, scale_factor, worker_type,
-                                            worker_state[worker_type],
-                                            new_worker_assignments)
+            per_worker_state = worker_state[worker_type]
+            assigned_worker_ids = per_worker_state['assigned_worker_ids']
+            current_job = 0
+            scale_factors = set([x[1] for x in scheduled_jobs[worker_type]])
+            scale_factors = sorted(scale_factors, reverse=True)
+
+            # Assign workers in order of decreasing scale factor to prioritize
+            # locality for multi-GPU jobs.
+            for current_scale_factor in scale_factors:
```

Comment:
What jobs are you trying to assign workers here? What jobs have already been allocated workers?

Reply:
So basically it's going to alternate between assigning workers for jobs that can extend their leases and then assigning for remaining jobs, but do this one scale factor at a time. Previously it would just extend all the leases it could regardless of scale factor and then assign workers for all remaining jobs - this can lead to distributed jobs not being placed on the same server.